shell/browser/extensions/electron_extension_loader.cc

```diff
@@ -41,6 +43,15 @@ std::pair<scoped_refptr<const Extension>, std::string> LoadUnpacked(
     return std::make_pair(nullptr, err);
   }
 
+  // remove _metadata folder. Otherwise, the following warning will be thrown
+  // Cannot load extension with file or directory name _metadata.
+  // Filenames starting with "_" are reserved for use by the system.
+  // see: https://bugs.chromium.org/p/chromium/issues/detail?id=377278
+  base::FilePath metadata_dir = extension_dir.Append(kMetadataFolder);
+  if (base::DirectoryExists(metadata_dir)) {
+    base::DeletePathRecursively(metadata_dir);
+  }
```

Comment:
I think we should determine why Chromium deems this necessary and possibly allow toggling of this behavior. Where in Chromium does this error get thrown and why?

I've seen apps loading Chrome extensions from user installations of Chrome. Potentially removing data does not seem desirable.

Reply:
👍  to what @samuelmaddock said, it's a big 🙅  to destructively delete data off disk for an operation that doesn't obviously do so ("load extension" does not imply "modify data on disk")

Reply:
The actual use of this directory is to verify the plugin's hash in case it is tampered with. see: https://stackoverflow.com/questions/2250163/how-do-i-directly-modify-a-google-chrome-extension-file-crx/4980410#4980410

Reason for error reporting: 
[electron/shell/browser/extensions/electron_extension_loader.cc#L45](https://github.com/electron/electron/blob/main/shell/browser/extensions/electron_extension_loader.cc#L45) -> [chromium/extensions/common/file_util.cc#L244](https://source.chromium.org/chromium/chromium/src/+/main:extensions/common/file_util.cc;l=244;drc=6efee78914660d0249e01aa403579476d8dc6a20) -> [chromium/extensions/common/file_util.cc#L302](https://source.chromium.org/chromium/chromium/src/+/main:extensions/common/file_util.cc;l=302;drc=6efee78914660d0249e01aa403579476d8dc6a20) -> [chromium/extensions/common/file_util.cc#L385-388](https://source.chromium.org/chromium/chromium/src/+/main:extensions/common/file_util.cc;l=385-388;drc=6efee78914660d0249e01aa403579476d8dc6a20)

There is a judgment in chromium:

```c++
// Some filenames are special and allowed to start with "_".
if (filename == kLocaleFolder || filename == kPlatformSpecificFolder ||
    filename == FILE_PATH_LITERAL("__MACOSX")) {
  continue;
}
```

The `_metadata` directory is missing here, which causes the warning.

On this point, I have submitted a PR to chromium and it is currently under review: https://chromium-review.googlesource.com/c/chromium/src/+/2959994/

Judging from the chromium team's response, this issue should be fixed.

But waiting for chromium to merge, and then for electron to change the chromium upstream hash it depends on, the whole process can be rather lengthy.

So currently we have 4 solutions:
1. Wait for chromium to update, then electron will update chromium's hash
2. Submit a temporary patch
3. rename _metadate to metadata(As long as there is no _ in front of the directory there will be no such warning)
4. remove _metadata folder

Reply:
We occasionally backport patches from Chromium to earlier versions of Electron, but generally only when there's a good reason to do so. I don't think removing a warning message qualifies, personally, but I don't mind especially if it does.

We roll from Chromium's main branch daily into a PR, and merge roughly weekly, so this change would likely make it into Electron 14.x or 15.x.

Reply:
So with this fix, is there any hope of merging it into electron 11 and other versions?

As well as the current pr, your suggestion is to remove this fix and wait for the chromium side to merge it, right?

Reply:
Electron 11 is currently accepting security fixes only, which this isn't. It would be theoretically possible to land this in 12 and 13 but I don't see a compelling reason to do so.

Reply:
Well, because it was a tough choice for us to upgrade electron, and my intention was actually to block this annoying warning on top of not upgrading electron.

What you said makes sense. But can we have a discussion for this pr under the issues that have discussed this issue? Look at the options within the community for this. (Personally I'd like to fight for it)

Personally, a warning that is annoying and doesn't have any way to be turned off is a pain.

Reply:
Let's reconsider once we have a final PR that doesn't modify the extension directory.

Reply:
So I looked into this and it turns out, to my surprise, Chrome actually removes the `_metadata` folder when loading unpacked extensions.

https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/extensions/unpacked_installer.cc;l=246-252;drc=c8a6933b40b57186a6b0682d0eafadb8f93da62e

I'm wondering if possibly one of the extension load flags could possibly disable the `_metadata` folder check https://source.chromium.org/chromium/chromium/src/+/main:extensions/common/extension.h;l=82;drc=c8a6933b40b57186a6b0682d0eafadb8f93da62e

Reply:
😮  hmmmmm, if Chrome does it then this might not be totally surprising.

Reply:
after I spoke with chromium, their response:

1. Why can't you load crx packages directly?
2. You seem to have access to the filesystem. Why can't the code you pointed to modify the unzipped package and remove the _metadata directory?

For now, it looks like we can delete the `_metadate` folder at the electron level

Also I have a question, why can't we load crx? Instead, we have to decompress the crx before we can load the extension, what is the reason for this?

Reply:
The reason for not loading CRX directly is that I wanted the API to be non-persistent. i.e. you should have to call `loadExtension()` on every boot. The reason for _that_ is that if you _remove_ a call to `loadExtension` from your app, the extension should stop loading. Previously the API was persistent, so if you called `loadExtension` _once_ then the extension would be loaded every boot forever more, which caused a lot of confusion.

Loading a CRX requires that you decompress the CRX to some location on disk. Chrome does this in the user-data directory, which makes sense because loading an extension in Chrome is persistent, i.e. the extension will be automatically loaded on next boot.

If we were to add support for loading a CRX, we should add it as a separate (set of) APIs that persists & makes that persistence obvious, perhaps e.g.

```js
if (!ses.isExtensionInstalled('some_extension_id')
  ses.installExtension('/path/to/some_extension_id.crx')
```

along with associated `ses.uninstallExtension` and so on.


Separately, given that Chrome deletes the _metadata directory, I think we're fine to do it too, since users may expect that.

Reply:
I'm talking to the chromium team about the possibility of adding new flags to chromium to do this. If there is any progress, I will give feedback here

Reply:
The chromium team did not reply, so here first do not do flags function